dandiapi/api/dandiset_migration.py

```diff
@@ -0,0 +1,75 @@
+"""
+Maps the identifiers of all Dandisets to the identifier listed in their metadata.
+
+Identifier cycles are not addressed.
+If you have Dandiset 000001 with {"identifier":000002} and Dandiset 000002
+with {"identifier":000001}, manual intervention is required.
+"""
+from dandiapi.api.models.dandiset import Dandiset
+
+# Track log messages instead of logs.appending so they can be output in rest response
+logs = []
+
+
+def get_new_identifier(dandiset):
+    metadata = dandiset.most_recent_version.metadata.metadata
+    if 'identifier' not in metadata:
+        logs.append(f'Dandiset {dandiset.identifier} does not specify an identifier')
+        return None
+    try:
+        identifier = int(metadata['identifier'])
+    except ValueError:
+        logs.append(
+            f'Dandiset {dandiset.identifier} has a bad metadata identifier {metadata["identifier"]}'
+        )
+        return None
+
+    if 0 > identifier or identifier > 999999:
+        logs.append(
+            f'Dandiset {dandiset.identifier} has a bad metadata identifier {metadata["identifier"]}'
+        )
+        return None
+
+    if dandiset.id == identifier:
+        logs.append(f'Dandiset {dandiset.identifier} already has the correct identifier')
+        return None
+
+    if Dandiset.objects.filter(id=identifier):
+        logs.append(
+            f'Dandiset {dandiset.identifier} cannot be copied because {identifier} already exists'
+        )
+        return None
+
+    return identifier
+
+
+def move(dandiset, new_id):
+    # Grab the old dandiset's relevant information
+    owners = dandiset.owners
+    versions = list(dandiset.versions.all())
+
+    # Create a new dandiset with the new identifier
+    new_dandiset = Dandiset(id=new_id)
```

Comment:
This is likely just due to my inexperience, but does specifying the `id` here have any adverse effects on the ordering/numbering of new dandisets? Put another way, what are the effects of creating a new dandiset normally, and how does specifying the `id` here change that?

Reply:
Django provides an auto-incrementing `id` field [for free](https://docs.djangoproject.com/en/3.1/topics/db/models/#automatic-primary-key-fields). If it's not specified, it just picks the next int in the sequence.

Regardless of whether you used the auto-incremented value or explicitly specified an id as we did here, creating an instance of the model does not do any collision checking. Even if dandiset `000007` already exists, there's no problem with calling `Dandiset(id=7)` (we use `id` to generate `identifier`, which is the 0-padded string). If, however, you try to call `save()` on that duplicate dandiset, a `Dandiset.IntegrityError` is thrown with some info on the underlying Postgres error (in this case, duplicate unique columns).

Technically, we should do some error handling here in case there is a collision. However, we already checked for an existing dandiset as the last step in `get_new_identifier`, which is good enough for this script. There are some problems with race conditions during simultaneous runs, but that's not a consideration for a script like this.